File: backend/services/post-rust-actix-web-fdm/src/common/api/invalid_user_name.rs
use crate::common::utils::error::*;

impl super::super::UserName {
    pub(in crate::common) fn user_name_empty() -> ErrorBody {
        ErrorBody {
            error: ErrBody {
                error: "USER_NAME_EMPTY".into(),
                reason: "User name cannot be empty".to_string(),
                message: "User name cannot be empty".to_string(),
            },
        }
    }
    pub(in crate::common) fn user_name_too_short() -> ErrorBody {
        ErrorBody {
            error: ErrBody {
                error: "USER_NAME_TOO_SHORT".into(),
                reason: "User name must be at least 3 characters".to_string(),
                message: "User name must be at least 3 characters".to_string(),
            },
        }
    }
    pub(in crate::common) fn user_name_too_long() -> ErrorBody {
        ErrorBody {
            error: ErrBody {
                error: "USER_NAME_TOO_LONG".into(),
                reason: "User name must be at most 20 characters".to_string(),
                message: "User name must be at most 20 characters".to_string(),
            },
        }
    }
    pub(in crate::common) fn user_name_contains_illegal_character() -> ErrorBody {
        ErrorBody {
            error: ErrBody {
                error: "USER_NAME_ILLEGAL".into(),
                reason: "user name contains illegal character".to_string(),
                message: "user name contains illegal character".to_string(),
            },
        }
    }
}

File: backend/services/post-kt-vertx-fdm/src/main/kotlin/me/jason5lee/post_kt_vertx_fdm/common/api/InvalidUserName.kt
package me.jason5lee.post_kt_vertx_fdm.common.api

import me.jason5lee.post_ktor_mongo_fdm.common.utils.Err
import me.jason5lee.post_ktor_mongo_fdm.common.utils.FailureBody

internal object InvalidUserName {
    internal val empty: FailureBody = FailureBody(
        error = Err(
            error = "USER_NAME_EMPTY",
            reason = "User name cannot be empty",
            message = "User name cannot be empty",
        )
    )
    internal val tooShort: FailureBody = FailureBody(
        error = Err(
            error = "USER_NAME_TOO_SHORT",
            reason = "User name must be at least 3 characters",
            message = "User name must be at least 3 characters",
        )
    )
    internal val tooLong: FailureBody = FailureBody(
        error = Err(
            error = "USER_NAME_TOO_LONG",
            reason = "User name must be at most 20 characters",
            message = "User name must be at most 20 characters",
        )
    )
    internal val containsIllegalCharacter: FailureBody = FailureBody(
        error = Err(
            error = "USER_NAME_ILLEGAL",
            reason = "User name can only contain letters, numbers, and underscores",
            message = "User name can only contain letters, numbers, and underscores",
        )
    )
}
